### PR TITLE
CI: Removing External Service Dependency in CI

### PR DESCRIPTION
**Pull request by ProKil**
Update the CI test flow with docker compose. 

Now we don't need any external service for the test workflow. (Except for codecov) So the whole CI test process is reproducible!

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed
- [x] Branch name follows `type/descript` (e.g. `feature/add-llm-agents`)
- [x] Ready for code review

### CODE REVIEW

File: sotopia/generation_utils/generate.py [hunk 1]
```diff
@@ -486,9 +486,9 @@ async def agenerate(
         input_values["format_instructions"] = output_parser.get_format_instructions()
 
     if structured_output:
-        assert (
-            model_name == "gpt-4o-2024-08-06"
-        ), "Structured output is only supported in gpt-4o-2024-08-06"
+        assert model_name == "gpt-4o-2024-08-06" or model_name.startswith(
+            "custom"
+        ), "Structured output is only supported in gpt-4o-2024-08-06 or custom models"
```
Comment (XuhuiZhou) [inline sotopia/generation_utils/generate.py:491]: do all the custom models support structured output?

File: sotopia/generation_utils/generate.py [hunk 2]
```diff
@@ -501,7 +501,15 @@ async def agenerate(
         instantiated_prompt = prompt_result.messages[0].content
         assert isinstance(output_parser, PydanticOutputParser)
         assert isinstance(instantiated_prompt, str)
-        client = OpenAI()
+        if model_name.startswith("custom"):
+            client = OpenAI(
```
Comment (XuhuiZhou) [inline sotopia/generation_utils/generate.py:505]: Can we make sure telling people that the custom model has to go through the openAI SDK?

**Review by XuhuiZhou — CHANGES_REQUESTED** [review]

File: sotopia/generation_utils/generate.py (hunk 1 above)
Comment (ProKil) [reply sotopia/generation_utils/generate.py:491]: No but this if branch is only triggered if they explicitly opt in for structured output. There is no way for us to verify whether an arbitrary custom model can use structured output.

**Review by XuhuiZhou — APPROVED** [review]
I think we need to add some doc about mypy and precommit 
for both the w container and wo container option.

Otherwise, I think things look good!

File: sotopia/generation_utils/generate.py (hunk 2 above)
Comment (ProKil) [reply sotopia/generation_utils/generate.py:505]: Done, added a doc. Closes #215

**Pull request merged**